xarray/core/netcdftimeindex.py

```diff
@@ -0,0 +1,180 @@
+    if resolution == 'hour':
+        start = date_type(parsed.year, parsed.month, parsed.day, parsed.hour)
+        return start, start + timedelta(hours=1, microseconds=-1)
+    if resolution == 'minute':
+        start = date_type(parsed.year, parsed.month, parsed.day, parsed.hour,
+                          parsed.minute)
+        return start, start + timedelta(minutes=1, microseconds=-1)
+    if resolution == 'second':
+        start = date_type(parsed.year, parsed.month, parsed.day, parsed.hour,
+                          parsed.minute, parsed.second)
+        return start, start + timedelta(seconds=1, microseconds=-1)
+    else:
+        raise KeyError
+
+
+def get_date_field(datetimes, field):
+    return [getattr(date, field) for date in datetimes]
+
+
+def _field_accessor(name, docstring=None):
+    def f(self):
+        return get_date_field(self._data, name)
+
+    f.__name__ = name
+    f.__doc__ = docstring
+    return property(f)
+
+
+def get_date_type(self):
+    return type(self._data[0])
+
+
+class NetCDFTimeIndex(pd.Index):
+    def __new__(cls, data):
+        result = object.__new__(cls)
+        result._data = np.array(data)
+        return result
+
+    year = _field_accessor('year', 'The year of the datetime')
+    month = _field_accessor('month', 'The month of the datetime')
+    day = _field_accessor('day', 'The days of the datetime')
+    hour = _field_accessor('hour', 'The hours of the datetime')
+    minute = _field_accessor('minute', 'The minutes of the datetime')
+    second = _field_accessor('second', 'The seconds of the datetime')
+    microsecond = _field_accessor('microsecond',
+                                  'The microseconds of the datetime')
+    date_type = property(get_date_type)
+
+    def _partial_date_slice(self, resolution, parsed,
+                            use_lhs=True, use_rhs=True):
+        start, end = _parsed_string_to_bounds(self.date_type, resolution,
+                                              parsed)
+        lhs_mask = (self._data >= start) if use_lhs else True
+        rhs_mask = (self._data <= end) if use_rhs else True
+        return (lhs_mask & rhs_mask).nonzero()[0]
+
+    def _get_string_slice(self, key, use_lhs=True, use_rhs=True):
+        parsed, resolution = _parse_iso8601_with_reso(self.date_type, key)
+        loc = self._partial_date_slice(resolution, parsed, use_lhs, use_rhs)
+        return loc
+
+    def get_loc(self, key, method=None, tolerance=None):
+        if isinstance(key, pd.compat.string_types):
+            result = self._get_string_slice(key)
+            # Prevents problem with __contains__ if key corresponds to only
+            # the first element in index (if we leave things as a list,
+            # np.any([0]) is False).
+            # Also coerces things to scalar coords in xarray if possible,
+            # which is consistent with the behavior with a DatetimeIndex.
+            if len(result) == 1:
+                return result[0]
```

Comment:
You're correct; DatetimeIndex doesn't do exactly this:
```
In [1]: import xarray as xr

In [2]: import pandas as pd

In [3]: rng = pd.date_range('2000-01-01', '2001-01-01', freq='M')

In [4]: da = xr.DataArray(rng, coords=[rng], dims=['time'])

In [5]: da.sel(time='2000-01')
Out[5]:
<xarray.DataArray (time: 1)>
array(['2000-01-31T00:00:00.000000000'], dtype='datetime64[ns]')
Coordinates:
  * time     (time) datetime64[ns] 2000-01-31

In [6]: da.sel(time='2000-01-31')
Out[6]:
<xarray.DataArray ()>
array(949276800000000000L, dtype='datetime64[ns]')
Coordinates:
    time     datetime64[ns] 2000-01-31
```
I'll try and simplify the logic here using your suggestion above.